### PR TITLE
Fix for #819 : Attribute Routing: Resolve Name + Token Replacement

### PR DESCRIPTION
**Pull request by kirthik**
@rynowak : Could you please help take a look?

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/ControllerActionDescriptorBuilder.cs [hunk 1]
```diff
@@ -80,6 +80,12 @@ public static IList<ControllerActionDescriptor> Build(ApplicationModel applicati
                             // for this action.
                             ReplaceAttributeRouteTokens(actionDescriptor, routeTemplateErrors);
 
+                            var attributeRouteInfo = actionDescriptor.AttributeRouteInfo;
+                            if (attributeRouteInfo.Name != null)
+                            {
+                                attributeRouteInfo.Name = attributeRouteInfo.Name + "_" + action.ActionName;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ControllerActionDescriptorBuilder.cs]: I think there's some confusion here and a wire got crossed. I was expecting this fix to run the token replacement code on the route name. I reread your email and I see what you're saying there, I must have missed that you were proposing something different.

So `Products_[action]` -> `Products_Edit`.

The problems with what you're doing here are twofold:
1). This doesn't account for the use of `[controller]` or `[area]` - both of which are legal
2). This is magic - you're taking something the user wrote, and turning it into something she didn't write.

**Comment by kirthik** [comment]
@rynowak - updated code. Could you please take a look?

**Comment by rynowak** [comment]
:shipit: conditionally - can you please try one thing - add a testscase which uses an unknown token in the route name, and **verify that the error message makes sense**.

If the error message doesn't make sense, let's fix that too.

**Comment by kirthik** [comment]
Here is the error message we see when unknown token is used - 
```
While processing template 'Products_[unknown]', a replacement value for the token 'unknown' could not be found. Available tokens: 'action, controller'.
```

Seems ok to me.

**Comment by rynowak** [comment]
Yep. Seems Ok.

**Comment by kirthik** [comment]
checked in - 9d2b1822d9e93bf74e3f1df8096b6bb2d6d43371

**Pull request closed without merging**